File: practices/practice4.1/src/ConcretePizzaBuilder.cpp
#include "ConcretePizzaBuilder.hpp"

ConcretePizzaBuilder::ConcretePizzaBuilder(const std::map<std::string, std::pair<double, bool>>& ingredients) : ingredients(ingredients) {
    pizza = std::make_unique<Pizza>(ingredients);
}

void ConcretePizzaBuilder::addIngredient(const std::string& ingredient, int count) {
    pizza->addIngredient(ingredient, ingredients[ingredient].first, ingredients[ingredient].second, count);
}

void ConcretePizzaBuilder::removeIngredient(const std::string& ingredient, int count) {
    pizza->removeIngredient(ingredient, count);
}

std::unique_ptr<Pizza> ConcretePizzaBuilder::getPizza() {
    std::unique_ptr<Pizza> oldPizza = std::move(pizza);
    pizza = std::make_unique<Pizza>(ingredients);
    return oldPizza;
}

File: homeworks/homework4.1/src/main/java/com/pizza/ConcretePizzaBuilder.java
package com.pizza;

import java.util.Map;

public class ConcretePizzaBuilder implements PizzaBuilder {
    private final Pizza pizza;
    private final Map<String, Ingredient> ingredients;

    public ConcretePizzaBuilder(Map<String, Ingredient> ingredients) {
        this.ingredients = ingredients;
        this.pizza = new Pizza(ingredients);
    }

    @Override
    public void addIngredient(String name, int count) {
        Ingredient ingredient = ingredients.get(name);
        if (ingredient != null) {
            pizza.addIngredient(name, ingredient.getPrice(), ingredient.isRequired(), count);
        }
    }

    @Override
    public void removeIngredient(String name, int count) {
        pizza.removeIngredient(name, count);
    }

    @Override
    public Pizza getPizza() {
        return pizza;
    }
}
